pkg/logicalbackup/logicalbackup.go

```diff
@@ -489,61 +489,79 @@ func (b *LogicalBackup) prepareTablesForPublication(conn *pgx.Conn) error {
 	// fetch all tables from the current publication, together with the information on whether we need to create
 	// replica identity full for them
 	type tableInfo struct {
-		oid                uint32
-		t                  message.Identifier
-		hasReplicaIdentity bool
+		oid             uint32
+		name            message.Identifier
+		hasPK           bool
+		replicaIdentity message.ReplicaIdentity
 	}
 	rows, err := conn.Query(`
 			select c.oid,
 				   n.nspname,
 				   c.relname,
-       			   CASE
-         			WHEN csr.oid IS NULL AND c.relreplident IN ('d', 'n') THEN true
-         			ELSE false END AS has_replica_identity
```

Comment:
It's better to avoid putting any logic into the query

Reply:
why? I think it's perfectly fine to make a decision write into this query:
```
select i.oid,
       i.nspname,
       i.relname,
       case when i.has_pk and i.replica_identity = 'n'
            then true
            else false
       end as set_replica_identity_to_default,
       case when not i.has_pk and i.replica_identity in ('n', 'd')
            then true
            else false
       end as set_replica_identitiy_to_full
from (select c.oid,
             n.nspname,
             c.relname,
             csr.oid is not null as has_pk,
             c.relreplident as replica_identity
      from pg_class c
             join pg_namespace n on n.oid = c.relnamespace
             join pg_publication_tables pub on (c.relname = pub.tablename and n.nspname = pub.schemaname)
             left join pg_constraint csr on (csr.conrelid = c.oid and csr.contype = 'p')
      where c.relkind = 'r'
        and pub.pubname = 'mypub') i
```
I'd even go further and do actual changes right here in a do block; however, that would lock multiple tables in an access exclusive lock, which is a sure recipe for a deadlock with some other users of those tables.

Reply:
I might be old-fashioned, but for me database is for storing data, logic should be done on the application side

Reply:
It's not about storing or not storing, it is always about how much logic are we storing inside the database (https://tapoueh.org/blog/2017/06/sql-and-business-logic/).

In our case it would have saved us some spilling of the database internal knowledge (replica identities) into the go code.

Reply:
we can come back to that topic later. as for now I want things work